main.c

```diff
@@ -612,6 +612,63 @@ apply_excludes (Pool       *pool,
   return excludes;
 }
 
+static void
+add_solvable_to_queue (const char *solvable,
+                       Pool       *pool,
+                       Queue      *selection,
+                       int         flags,
+                       Queue      *solvq)
+{
+  selection_make (pool, selection, solvable, flags);
+  g_auto(Queue) q;
+  queue_init (&q);
+  selection_solvables (pool, selection, &q);
+  if (!q.count)
+    {
+      g_warning ("Nothing matches '%s'", solvable);
+      return;
+    }
+  pool_best_solvables (pool, &q, 0);
+  for (int j = 0; j < q.count; j++)
+    queue_push (solvq, q.elements[j]);
+}
+
+static void
+add_solvables_from_file_to_queue (const char *filename,
+                                  Pool       *pool,
+                                  Queue      *selection,
+                                  int         flags,
+                                  Queue      *solvq)
+{
+  g_autoptr(GError) err = NULL;
+  GIOChannel *ch = g_io_channel_new_file (filename, "r", &err);
+  if (ch == NULL)
+  {
+    g_printerr ("Unable to open file '%s': %s\n", filename, err->message);
+    exit (EXIT_FAILURE);
```

Comment:
what I meant is that function should never do `exit()`. You should set error and return FALSE. So you should make function to return bool